File: src/year2021/day25/day25.rs
//! [Day 25: Sea Cucumber](https://adventofcode.com/2021/day/25)

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (i32, aoc::Christmas) {
    let data = data.lines().collect::<Vec<&str>>();

    let nx = data[0].len();
    let ny = data.len();
    let mut grid = vec![vec![b'.'; nx]; ny];

    // load the grid
    for (y, line) in data.iter().enumerate() {
        for (x, col) in line.bytes().enumerate() {
            grid[y][x] = col;
        }
    }

    // move the sea cucumbers
    let mut step = 1;
    while do_move(&mut grid) {
        step += 1;
    }

    (step, aoc::CHRISTMAS)
}

fn do_move(grid: &mut [Vec<u8>]) -> bool {
    let mut moved = false;

    let nx = grid[0].len();
    let ny = grid.len();

    // don't move blocked sea cucumbers
    for y in 0..ny {
        for x in 0..nx {
            if grid[y][x] == b'>' && grid[y][x] == grid[y][(x + 1) % nx] {
                grid[y][x] = b'H';
            }
            if grid[y][x] == b'v' && grid[y][x] == grid[(y + 1) % ny][x] {
                grid[y][x] = b'V';
            }
        }
    }

    // During a single step, the east-facing herd moves first,
    for line in grid.iter_mut() {
        for x in 0..nx {
            if line[x] == b'>' && line[(x + 1) % nx] == b'.' {
                line[(x + 1) % nx] = b'H';
                line[x] = b'.';
                moved = true;
            }
        }
    }

    // then the south-facing herd moves.
    for y in 0..ny {
        for x in 0..nx {
            let c = grid[y][x];
            if c == b'v' && grid[(y + 1) % ny][x] == b'.' {
                grid[(y + 1) % ny][x] = b'V';
                grid[y][x] = b'.';
                moved = true;
            }
        }
    }

    // restore blocked and moving sea cucumbers
    for line in grid {
        for val in line {
            match val {
                b'H' => *val = b'>',
                b'V' => *val = b'v',
                _ => (),
            }
        }
    }

    // indicate if any sea cucumbers moved
    moved
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn part1() {
        assert_eq!(solve(TEST_INPUT).0, 58);
    }
}

File: src/year2021/day25/day25.py
#!/usr/bin/env python3

# Day 25: Sea Cucumber
# https://adventofcode.com/2021/day/25

import atexit
import sys
import time

data = [line.strip() for line in open("input.txt" if len(sys.argv) == 1 else sys.argv[1])]
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


ny = len(data)
nx = len(data[0])

grid = [["." for _ in range(nx)] for _ in range(ny)]
for y, line in enumerate(data):
    for x, c in enumerate(line):
        grid[y][x] = c


def prt():
    print("\n".join(["".join(line) for line in grid]))
    print()


def move():
    moved = False

    # don't move blocked sea cucumbers
    for y in range(ny):
        for x in range(nx):
            if grid[y][x] == grid[y][(x + 1) % nx] == ">":
                grid[y][x] = "H"
            if grid[y][x] == grid[(y + 1) % ny][x] == "v":
                grid[y][x] = "V"

    # « During a single step, the east-facing herd moves first,
    for y in range(ny):
        for x in range(nx):
            c = grid[y][x]
            if c == ">" and grid[y][(x + 1) % nx] == ".":
                grid[y][(x + 1) % nx] = "H"
                grid[y][x] = "."
                moved = True

    # then the south-facing herd moves. »
    for y in range(ny):
        for x in range(nx):
            c = grid[y][x]
            if c == "v" and grid[(y + 1) % ny][x] == ".":
                grid[(y + 1) % ny][x] = "V"
                grid[y][x] = "."
                moved = True

    # restore blocked and moving sea cucumbers
    for y in range(ny):
        for x in range(nx):
            c = grid[y][x]
            if c == "H":
                grid[y][x] = ">"
            elif c == "V":
                grid[y][x] = "v"

    return moved


step = 1
while move():
    step += 1
print(step)
